### PR TITLE
[Feature] Support triggering compaction for a specific partition manually

### LINKED ISSUE #7518: [Enhancement] Support triggering compaction for specific partition manually
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Sometimes there are large amounts of data marked for deletion for a specific table/partition,  which cannot be removed from the disk in time, so we need to increase the priority of compaction for the specific partition to clean data as quickly as possible.

### Solution

Add statement to trigger cumulative or base compaction for all replicas of a specified partition.

Grammar:
```
ADMIN COMPACT TABLE table_name PARTITION partition_name WHERE TYPE='BASE/CUMULATIVE'
```

Explain:
* This statement only means that the system attempts to submit a compaction task for each replica under the specified partition to compaction thread pool, and it is not guaranteed to be successful. It only increases the priority of compaction for the specific partition.
* This statement only supports triggering compaction task for a single partition of the table at a time.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by weizuo93**
## Proposed changes

Close related #7518 

Add statement to trigger cumulative or base compaction for a specified partition.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #7518 ) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

File: be/src/olap/olap_server.cpp [hunk 1]
```diff
@@ -551,4 +537,51 @@ void StorageEngine::_pop_tablet_from_submitted_compaction(TabletSharedPtr tablet
...
+        // reset compaction
+        tablet->reset_compaction(compaction_type);
+        _pop_tablet_from_submitted_compaction(tablet, compaction_type);
+        return Status::InternalError(strings::Substitute(
+                "failed to prepare compaction task and calculate permits, tablet_id=$0, compaction_type=$1.",
+                tablet->tablet_id(), compaction_type));
+    }
+}
+
+Status StorageEngine::submit_table_compaction_task(TabletSharedPtr tablet, CompactionType compaction_type) {
```
Comment (morningman) [inline be/src/olap/olap_server.cpp:578]:
```suggestion
Status StorageEngine::submit_compaction_task(TabletSharedPtr tablet, CompactionType compaction_type) {
```

File: be/src/agent/task_worker_pool.cpp [hunk 2]
```diff
@@ -1650,4 +1654,73 @@ void TaskWorkerPool::_random_sleep(int second) {
...
+        if (tablet_ptr != nullptr) {
+            auto data_dir = tablet_ptr->data_dir();
+            if (!tablet_ptr->can_do_compaction(data_dir->path_hash(), compaction_type)) {
+                LOG(WARNING) << "can not do compaction: " << tablet_ptr->tablet_id()
+                             << ", compaction type: " << compaction_type;
+                _remove_task_info(agent_task_req.task_type, agent_task_req.signature);
+                continue;
+            }
+
+            if (compaction_type == CompactionType::BASE_COMPACTION) {
```
Comment (morningman) [inline be/src/agent/task_worker_pool.cpp:1697]: Is it necessary to check lock here?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 3]
```diff
@@ -7257,4 +7259,46 @@ public static boolean isStoredTableNamesLowerCase() {
     public static boolean isTableNamesCaseInsensitive() {
         return GlobalVariable.lowerCaseTableNames == 2;
     }
+
+    public void compactTable(AdminCompactTableStmt stmt) throws DdlException {
+        String dbName = stmt.getDbName();
+        String tableName = stmt.getTblName();
+
+        String type = stmt.getCompactionType();
+        if (type == null || (!type.equals("base") && !type.equals("cumulative"))) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:7268]: This check should be done in analysis phase

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 4]
```diff
@@ -7257,4 +7259,46 @@ public static boolean isStoredTableNamesLowerCase() {
...
+            throw new DdlException("compaction type should be [BASE] or [CUMULATIVE]");
+        }
+
+        Database db = this.getDbOrDdlException(dbName);
+        OlapTable olapTable = db.getOlapTableOrDdlException(tableName);
+
+        olapTable.writeLock();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:7275]: readLock is enough

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 5]
```diff
@@ -7257,4 +7259,46 @@ public static boolean isStoredTableNamesLowerCase() {
...
+                            CompactionTask compactionTask = new CompactionTask(replica.getBackendId(), db.getId(),
+                                    olapTable.getId(), partition.getId(), idx.getId(), tablet.getId(),
+                                    olapTable.getSchemaHashByIndexId(idx.getId()), type, 5000);
+                            batchTask.addTask(compactionTask);
+                        }
+                    }
+                } // indices
+            }
+            // send task immediately
+            AgentTaskExecutor.submit(batchTask);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:7299]: submit task outside the lock

File: gensrc/thrift/AgentService.thrift [hunk 6]
```diff
@@ -160,6 +160,13 @@ struct TCloneReq {
     10: optional i32 timeout_s;
 }
 
+struct TCompactionReq {
+    1: required Types.TTabletId tablet_id
```
Comment (morningman) [inline gensrc/thrift/AgentService.thrift:164]: use optional for all fields

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 5 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:7299]: Timeout for 5000 seconds?
I think the timeout can be calculated by BE.

File: docs/en/sql-reference/sql-statements/Administration/ADMIN COMPACT.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+---
```
Comment (morningman) [inline docs/en/sql-reference/sql-statements/Administration/ADMIN COMPACT.md:1]: New doc need to be added to the `docs/.vuepress/sidebar/en.js` and `zh-CN.js`

File: be/src/olap/olap_server.cpp [hunk 8]
```diff
@@ -551,4 +537,51 @@ void StorageEngine::_pop_tablet_from_submitted_compaction(TabletSharedPtr tablet
     }
 }
 
+Status StorageEngine::_submit_compaction_task(TabletSharedPtr tablet, CompactionType compaction_type) {
+    bool already_exist = _push_tablet_into_submitted_compaction(tablet, compaction_type);
+    if (already_exist) {
+        return Status::InternalError(strings::Substitute(
```
Comment (liutang123) [inline be/src/olap/olap_server.cpp:543]: Is `Status::AlreadyExist` more appropriate?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 3 above)
Comment (weizuo93) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java]: > This check should be done in analysis phase

OK, thank you.

File: be/src/agent/task_worker_pool.cpp (hunk 2 above)
Comment (weizuo93) [reply be/src/agent/task_worker_pool.cpp]: > Is it necessary to check lock here?

OK.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 4 above)
Comment (weizuo93) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java]: > readLock is enough

OK, thank you.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 5 above)
Comment (weizuo93) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java]: > submit task outside the lock

Done.

File: docs/en/sql-reference/sql-statements/Administration/ADMIN COMPACT.md (hunk 7 above)
Comment (weizuo93) [reply docs/en/sql-reference/sql-statements/Administration/ADMIN COMPACT.md:1]: > New doc need to be added to the `docs/.vuepress/sidebar/en.js` and `zh-CN.js`

Done.

File: be/src/olap/olap_server.cpp (hunk 8 above)
Comment (weizuo93) [reply be/src/olap/olap_server.cpp]: > Is `Status::AlreadyExist` more appropriate?

OK, thank you.

File: fe/fe-core/src/main/cup/sql_parser.cup [hunk 9]
```diff
@@ -5226,6 +5226,10 @@ admin_stmt ::=
     {:
         RESULT = new AdminCancelRepairTableStmt(table_ref);
     :}
+    | KW_ADMIN KW_COMPACT KW_TABLE base_table_ref:table_ref opt_wild_where
```
Comment (morningman) [inline fe/fe-core/src/main/cup/sql_parser.cup:5229]: Add `KW_COMPACT` to the `keyword ::=` region

Comment (weizuo93) [reply fe/fe-core/src/main/cup/sql_parser.cup:5229]: > Add `KW_COMPACT` to the `keyword ::=` region

Done.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**